File: cli-ruby/lib/input_commands_processor.rb
PLACE_COMMAND_NAME = 'PLACE'
PLACE_COMMAND_REGEX = /(\bPLACE\b) \d+,\d+,((\bNORTH\b)|(\bEAST\b)|(\bSOUTH\b)|(\bWEST\b))/
OTHER_COMMANDS_REGEX = /(\bMOVE\b)|(\bLEFT\b)|(\bRIGHT\b)|(\bREPORT\b)/

class InputCommandsProcessor
  def initialize(input)
    @input = input
  end

  def process
    return [] if @input.nil? || @input.empty?

    commands = @input.upcase.split(/(\n)/)
    valid_commands = filter_valid_commands(commands)
    build_commands(valid_commands)
  end

  private

  def filter_valid_commands(commands)
    return [] if (commands.nil? || commands.empty?)

    commands.select { |command| valid_command?(command) }
  end

  def valid_command?(command)
    return false if command.nil? || command.empty?

    if command.include?(PLACE_COMMAND_NAME)
      command.match?(PLACE_COMMAND_REGEX)
    else
      command.match?(OTHER_COMMANDS_REGEX)
    end
  end

  def build_commands(valid_commands)
    return [] if valid_commands.nil? || valid_commands.empty?

    commands = []
    valid_commands.each do |command_str|
      if command_str.include?(PLACE_COMMAND_NAME)
        command = build_place_command(command_str)
      else
        command = build_simple_command(command_str)
      end
      commands.push(command)
    end
    commands
  end

  def build_place_command(command)
    parts_of_command = command.split(' ')
    name = parts_of_command[0]
    params = parts_of_command[1].split(',')

    {
      name: name.to_sym,
      params: {
        x: params[0].to_i,
        y: params[1].to_i,
        facing: params[2].to_sym
      }
    }
  end

  def build_simple_command(command_str)
    {
      name: command_str.to_sym
    }
  end
end

File: web-app/src/input_commands_processor.js
import {COMMANDS, COMMANDS_REGEX_VALIDATION} from "./constants.js";

class _InputCommandsProcessor {
    constructor(commands) {
        this.commandsInput = commands;
    }

    process() {
        if (!this.commandsInput) return [];
        const commands = this.commandsInput.toUpperCase().split(/(\n)/);
        const validsCommands = this.filterValidCommands(commands)
        return this.buildCommands(validsCommands)
    }

    filterValidCommands(commands) {
        if (!commands) return [];
        return commands.filter((command) => this.isValidCommand(command));
    }

    buildCommands(commands) {
        if (!commands) return [];
        const buildedCommands = []
        commands.forEach(command => {
            const isPlaceCommand = command.includes(COMMANDS.place);
            const buildedCommand = isPlaceCommand ? this.buildPlaceCommand(command) : this.buildMovimentCommand(command)
            buildedCommands.push(buildedCommand)
        })
        return buildedCommands;
    }

    isValidCommand(command) {
        const regexValidation = COMMANDS_REGEX_VALIDATION[command] || COMMANDS_REGEX_VALIDATION.PLACE
        return regexValidation.test(command);
    }

    buildPlaceCommand(command) {
        const partsOfCommand = command.split(' ')
        const name = partsOfCommand[0]
        const params = partsOfCommand[1].split(',')
        return {
            name: name,
            params: {
                x: parseInt(params[0]),
                y: parseInt(params[1]),
                facing: params[2]
            }
        }
    }

    buildMovimentCommand(command) {
        return {name: command}
    }
}

const InputCommandsProcessor = (commands) => {
    const inputCommandsProcessor = new _InputCommandsProcessor(commands);

    return {
        process: () => inputCommandsProcessor.process(),
    }
}

export default InputCommandsProcessor;